packages/babel-helpers/src/helpers.js

```diff
@@ -1973,3 +1973,122 @@ helpers.wrapRegExp = helper("7.2.6")`
     return _wrapRegExp.apply(this, arguments);
   }
 `;
+helpers.slice = helper("7.6.0")`
+  import isNativeFunction from 'isNativeFunction';
+  function isString(value) {
+    return typeof value === "string" || (
+        typeof value === "object" && Object.prototype.toString.call(value) === "[object String]"
+    )
+  }
+
+  function toLength(value) {
+    var len = toInteger(value);
+    if (len < 0) {
+      return 0;
+    }
+    return Math.min(len, 9007199254740991);
+  }
+
+  function toInteger(value) {
+    const number = +value;
+    if (isNaN(number)) {
+      return 0;
+    }
+    if (number === 0 || number === Infinity || number === -Infinity) {
+      return number;
+    }
+    return number > 0 ? Math.floor(number) : -Math.floor(-number);
+  }
+
+  export default function slice(t, begin, end, step) {
+    // The default value for the step argument is 1
+    if (step === void 0) {
+      step = 1;
+    }
+    step = toInteger(step);
+    if (step === 0 || step === Infinity || step === -Infinity) {
+      // todo: the proposal does not specify details about these cases
+      // https://github.com/tc39/proposal-slice-notation/issues/28
+      throw new Error("Slice step can not be zero, NaN of Infinity");
+    }
+    var tIsString = isString(t);
+    var tIsArray = Array.isArray(t);
+    if (
```

Comment:
I think this is too complicated for a proposal that doesn't have defined semantics yet.

Reply:
> This notation can be used for slice operations on primitives like String and any object that provides indexed access using [[Get]] like Array and TypedArray.

It is used to limit the slicee to string, array and typed array. Surely we can debate about whether we should loose the control here and allow any fancy objects as long as they implement indexed access.

Reply:
That matches array-likes, too, eg `{ length: 1, 0: 'test' }`

Reply:
> That matches array-likes, too, eg { length: 1, 0: 'test' }.

Yes, array-like objects is supported.

```js
const obj = { 0: 'a', 1: 'b', 2: 'c', 3: 'd', length: 4 };
obj[1:3];
// → ['b', 'c']
```

But we should reject the following Proxy exotic object
```js
new Proxy({} , { 
  get(target, prop) {
    if (prop === "length") { return 1 }
    if (prop == 0) { return prop }
    return undefined
  }
});
```

Reply:
> any object that provides indexed access using [[Get]]

I think that this means that proxies should be allowed. For example, when using a properly defined proxy, `[].slice` works:

```js
var obj = new Proxy({} , { 
  get(target, prop) {
    if (prop === "length") { return 1 }
    if (prop == 0) { return 3 }
  },
  has(target, prop) {
    return prop == 0;
  }
});

[].slice.call(obj); // [3]
```

cc @gsathya What should it do here?